### DISCUSSION TITLE
Prevent focuising node on update

### DISCUSSION THREAD

**Discussion by levi-putna** [ueberdosis/tiptap#4832] [General Questions]
**Background**

I'm working on a web editor using Tiptap. I've developed an external attributes panel that enables me to modify the attributes (padding, margin, background colour, etc.) of the selected block node. This panel includes several form elements, such as text fields, that correspond to the attributes and allow the values to be changed.

In this setup, I'm utilising Next.js with a client React component rendering the editor and other UI, then updating the corresponding block attribute when the value changes in the input or other form elements that are linked to the block attribute.

**Problem**

I'm encountering an issue where executing the command to update the attribute causes the editor to shift focus away from the input field and back to the editor node with the updated attribute. This can be particularly annoying if you need to type multiple values into a field and it loses focus after the first key is entered. For example, trying to add padding of 10, would loose focus after the 1 is entered. 

**Question**

Could you advise me on how to update the node's attributes without it regaining focus? I have been looking at overriding the ProseMirror appendTransaction function to look for a 'preventFocus' meta and avoid changing focus, but not having much luck.

```
import { Plugin } from 'prosemirror-state';

export const preventFocusPlugin = new Plugin({
    appendTransaction(transactions, oldState, newState) {
        if (transactions.some(tr => tr.getMeta('preventFocus'))) {
            // Create a new transaction that maintains the old selection
            const selectionTr = newState.tr;
            selectionTr.setSelection(oldState.selection);
            return selectionTr;
        }
    }
});
```

Here is the `updateAttribute` command that is hijacking focus. 

```
updateAttribute:
    (attributeName, attributeValue) =>
         ({ tr, state, dispatch }) => {
            const { selection } = state;
            const { from, to } = selection;

            tr.doc.nodesBetween(from, to, (node, pos) => {
                if (node.type.name === "section") {
                   const updatedAttrs = { ...node.attrs, [attributeName]: attributeValue };
                   tr.setMeta("preventFocus", true); // Set a custom meta flag to indicate this transaction shouldn't set focus

                   tr.setNodeMarkup( pos, null, updatedAttrs );
               }
           });

           if (dispatch) {
               dispatch(tr);
           }

       return tr;
},
```

Here is a screenshot of the editor for context. 

<img width="1083" alt="Screenshot 2024-01-26 at 6 58 22 pm" src="https://github.com/ueberdosis/tiptap/assets/582019/4816392c-a269-491b-8564-329b73e8ede0">

**Comment by kongweiying2** [comment]
I don't have a solution for you but I have a variant of this problem as well. Basically, when updateAttributes is called, any selection made inside a node is lost.

**Reply by kongweiying2** [reply]
@levi-putna Could you share your approach for creating a custom background colour? Completely different problem to your original post but I'm trying to achieve this too.
